homeassistant/components/habitica/calendar.py

```diff
@@ -0,0 +1,210 @@
+from .types import HabiticaTaskType
+from .util import build_rrule, get_recurrence_rule
+
+
+class HabiticaCalendar(StrEnum):
+    """Habitica calendars."""
+
+    DAILIES = "dailys"
+    TODOS = "todos"
+
+
+async def async_setup_entry(
+    hass: HomeAssistant,
+    config_entry: HabiticaConfigEntry,
+    async_add_entities: AddEntitiesCallback,
+) -> None:
+    """Set up the calendar platform."""
+    coordinator = config_entry.runtime_data
+
+    async_add_entities(
+        [
+            HabiticaTodosCalendarEntity(coordinator),
+            HabiticaDailiesCalendarEntity(coordinator),
+        ]
+    )
+
+
+class HabiticaCalendarEntity(HabiticaBase, CalendarEntity):
+    """Base Habitica calendar entity."""
+
+    def __init__(
+        self,
+        coordinator: HabiticaDataUpdateCoordinator,
+    ) -> None:
+        """Initialize calendar entity."""
+        super().__init__(coordinator, self.entity_description)
+
+
+class HabiticaTodosCalendarEntity(HabiticaCalendarEntity):
+    """Habitica todos calendar entity."""
+
+    entity_description = CalendarEntityDescription(
+        key=HabiticaCalendar.TODOS,
+        translation_key=HabiticaCalendar.TODOS,
+    )
+
+    def dated_todos(self) -> list[CalendarEvent]:
+        """Get all dated todos."""
+
+        return sorted(
+            [
+                CalendarEvent(
+                    start=start,
+                    end=start + timedelta(days=1),
+                    summary=task["text"],
+                    description=task["notes"],
+                    uid=task["id"],
+                )
+                for task in self.coordinator.data.tasks
+                if task["type"] == HabiticaTaskType.TODO
+                and not task["completed"]
+                and task.get("date")
+                and (
+                    start := dt_util.as_local(
+                        datetime.fromisoformat(task["date"])
+                    ).date()
+                )
+            ],
+            key=lambda event: (
+                event.start,
+                self.coordinator.data.user["tasksOrder"]["todos"].index(event.uid),
+            ),
+        )
+
+    @property
+    def event(self) -> CalendarEvent | None:
+        """Return the current or next upcoming event."""
+
+        return next(
+            (
+                e
+                for e in self.dated_todos()
+                if dt_util.start_of_local_day(e.start)
+                <= (now := dt_util.now())
+                < dt_util.start_of_local_day(e.end)
+                or dt_util.start_of_local_day(e.start) > now
+            ),
+            None,
+        )
+
+    async def async_get_events(
+        self, hass: HomeAssistant, start_date: datetime, end_date: datetime
+    ) -> list[CalendarEvent]:
+        """Return calendar events within a datetime range."""
+
+        return [
+            event
+            for event in self.dated_todos()
+            if dt_util.start_of_local_day(event.start) < end_date
```

Comment:
Can this leave the start/end date of the event alone? Either this is not needed or if it is needed then it seems odd to return events with dates outside of the range requested.

Reply:
Not sure what you mean, I am filtering the events that are within the requested range (`self.dated_todos` returns all events). I could maybe let the dated_todos method do the filtering. Would that be better?

Reply:
moved the filtering to the `dated_todos` method.